pkg/controller/gkenetworkparamset/gkenetworkparamset_controller.go

```diff
@@ -466,3 +477,37 @@ func (c *Controller) updateGKENetworkParamSetStatus(ctx context.Context, params
 	}
 	return nil
 }
+
+// syncDefaultPodRanges updates the "default" params with new Pod range if needed
+func (c *Controller) syncDefaultPodRanges(ctx context.Context, params *networkv1.GKENetworkParamSet) error {
+	if params.Name != networkv1.DefaultPodNetworkName {
+		return nil
+	}
+	originalRangeNames := params.Spec.PodIPv4Ranges.RangeNames
+	rangeNames := make([]string, len(originalRangeNames))
+	copy(rangeNames, originalRangeNames)
+
+	k8sNodesList, err := c.nodeLister.List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("failed to list node from cache: %v", err)
+	}
+	for _, n := range k8sNodesList {
+		for k, v := range n.Labels {
+			if k != networkv1.NodePoolPodRangeLabelPrefix {
+				continue
+			}
+			seen := false
+			for _, r := range originalRangeNames {
+				if r == v {
+					seen = true
+					break
+				}
+			}
+			if !seen {
+				rangeNames = append(rangeNames, v)
+			}
+		}
+	}
+	params.Spec.PodIPv4Ranges.RangeNames = rangeNames
+	return nil
+}
```

Comment:
If you are going through all the nodes here, you might as well handle the case where some range in the 'originalRangeNames' is no longer in use. I would change the structure of this loop to handle this as follows:

```
rangeNameMap = make(map[string], bool)
for all nodes with NodePoolPodRangeLabel {
   if podRangeName not in rangeNameMap {
        add range name to map
   }
}
params.Spec.PodIPv4Ranges.RangeNames = convert map to slice.
```

Reply:
Done, include the case if same range is not longer in use. When the node is adding/updating with new pod range that's not in the current default paramset, or when the node is removed, the node informer event handler adds the default paramset into the work queue and the reconciliation calls this sync function, the sync will loop through the cached node list to update the PodIPv4Ranges